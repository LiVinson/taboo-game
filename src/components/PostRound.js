import React from "react"

export default class PostRound extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      cards: props.cardsPlayed
    }

    this.determineCardsStatus = this.determineCardsStatus.bind(this)
  }



  changeCardStatus(newStatus, cardIndex) {
    //finds the word to update, then updates the status

    this.setState((state) => {
        const updatedCards = [...state]
        updatedCards[cardIndex].status = newStatus
        return {
            cards: updatedCards
        }
    })
  }

  confirmCards() {
      console.log("cards Confirmed")
      this.props.confirmRoundEnd(this.state)
  }

  render() {
    const { cards } = this.state
    const correctCards = cards.filter(card => card.status==="correct")
    const skippedCards = cards.filter(card => card.status==="skipped")
    const noneCards = cards.filter(card => card.status==="none")
    return (
      <div>
        <h1>Post Round</h1>
        <h3>Correct Cards</h3>
        <ul>
          {correctCards.map((card, index) => (
            <li key={card.word}>
              <p>{card.word}</p>
              <button
                onClick={() =>
                  this.changeCardStatus("skipped", index)
                }
              >
                Move to Skipped
              </button>
              <button
                onClick={() =>
                  this.changeCardStatus("none", index)
                }
              >
                Move to No Status
              </button>
            </li>
          ))}
        </ul>
        <h3>Skipped Cards</h3>
        <ul>
          {skippedCards.map((card) => (
            <li key={card.word}>
              <p>{card.word}</p>
              <button
                onClick={() =>
                  this.changeCardStatus("skipped", "correct", card.word)
                }
              >
                Move to Correct
              </button>
              <button
                onClick={() =>
                  this.changeCardStatus("skipped", "none", card.word)
                }
              >
                Move to No Status
              </button>
            </li>
          ))}
        </ul>
        <h3>No Status Cards</h3>
        <ul>
          {noneCards.map((card, index) => (
            <li key={card.word}>
              <p>{card.word}</p>
              <button
                onClick={() =>
                  this.changeCardStatus("correct", index)
                }
              >
                Move to Correct
              </button>
              <button
                onClick={() =>
                  this.changeCardStatus("skipped", card.word)
                }
              >
                Move to Skipped
              </button>
            </li>
          ))}
        </ul>
        {this.props.isWatcher 
            ? <button
                onClick={this.confirmCards}>Confirm Cards</button>
            : null }
        
      </div>
    )
  }
}
