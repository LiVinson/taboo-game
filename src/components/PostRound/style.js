import styled from 'styled-components'

//Adds margin to each card component & button in the entire page, except the last one.
export const StyledPostRound = styled.div`
	& > *:not(:last-child){
		margin-bottom: 2rem;
	}
`


