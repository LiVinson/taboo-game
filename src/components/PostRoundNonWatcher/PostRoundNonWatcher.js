import React from 'react'
import PropTypes from 'prop-types'
import { TabooCard } from 'components/shared/TabooCard'

const PostRoundNonWatcher = ({ children }) => {
	return <TabooCard tabooWord={'Cards Played'}>{children}</TabooCard>
}

PostRoundNonWatcher.propTypes = {
	children: PropTypes.node.isRequired
}

export default PostRoundNonWatcher
