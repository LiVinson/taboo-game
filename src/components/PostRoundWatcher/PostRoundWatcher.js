import React from 'react'
import PropTypes from 'prop-types'
import InstructionsText from 'components/shared/InstructionsText'
import CardsPlayed from 'components/CardsPlayed'
import { LargeButton } from 'components/shared/Button'
import Pending from 'components/shared/Pending'
import { ErrorMessage } from 'components/shared/FeedbackMessage'
import KeyWord from 'components/shared/KeyWord'

//Used for each of 3 status types to create buttons to change card status to the other 2 statuses
export const generateCardsPlayedButtonInfo = (statusArray, status, statusSelected, cb, isPending) => {
	const buttonInfo = statusArray
		.filter((cardType) => cardType !== status)
		.map((cardType) => {
			const button = { className: 'button' }
			button.text = cardType
			button.disabled = statusSelected.length < 1 || isPending ? true : false
			button.onClick = () => cb(status, cardType)
			return button
		})

	return buttonInfo
}

const PostRoundWatcher = ({
	cardStatuses,
	cardsPlayed,
	selection,
	handleCardSelection,
	selectedCards,
	updateSelectedCard,
	confirmRoundEnd,
	isPending,
	error,
}) => {
	
	//Loops over each card status, and returns a Taboo card with 2 buttons.
	//Filters cards played based on the status. Information needed to generate buttons,
	//select a card in the list, and change button status passed in
	// Then separates cards played in round based on current card status
	return (
		<React.Fragment>
			<InstructionsText align="center">
				Review the cards from this round below. Select the word and click the button to change it's status.
				Select <KeyWord>Confirm!</KeyWord> when done to calculate scores.
			</InstructionsText>
			{cardStatuses.map((status, index) => (
				<CardsPlayed
					// Only pass down error message if it is for this Card status. Indicated by error message include status type
					error={
						error.cards &&
						cardsPlayed.some(
							(card) => card.status === status && error.cards.includes(card.index.toString())
						)
							? error.cards
							: null
					}
					key={status}
					status={cardStatuses[index]}
					selected={selectedCards[index]}
					cardList={cardsPlayed.filter((card) => card.status === status)}
					handleChange={handleCardSelection}
					buttonInfo={generateCardsPlayedButtonInfo(
						cardStatuses,
						status,
						selectedCards[index], //selectedCards array order is same as status array order, so indexes correspond
						updateSelectedCard,
						isPending.cards //true when in progress of updating status
					)}
				/>
			))}
			<LargeButton text="Confirm!" disabled={isPending.round} onClick={confirmRoundEnd} />
			{isPending.round ? <Pending speed={300} message={'Updating scores'} /> : null}
			{error.round && <ErrorMessage error={error.round} />}
		</React.Fragment>
	)
}

PostRoundWatcher.propTypes = {
	cardStatuses: PropTypes.array.isRequired,
	cardsPlayed: PropTypes.array.isRequired,
	handleCardSelection: PropTypes.func.isRequired,
	selectedCards: PropTypes.array.isRequired,
	updateSelectedCard: PropTypes.func.isRequired,
	confirmRoundEnd: PropTypes.func.isRequired,
	isPending: PropTypes.object.isRequired,
	error: PropTypes.object.isRequired,
}

export default PostRoundWatcher
