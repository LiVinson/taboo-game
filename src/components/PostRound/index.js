import PostRound from "./PostRound"

export default PostRound;