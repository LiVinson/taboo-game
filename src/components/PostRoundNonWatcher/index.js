import PostRoundNonWatcher from "./PostRoundNonWatcher"

export default PostRoundNonWatcher