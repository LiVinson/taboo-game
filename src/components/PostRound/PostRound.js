import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import PostRoundNonWatcher from 'components/PostRoundNonWatcher'
import PostRoundWatcher from 'components/PostRoundWatcher'
import { FilteredTabooList } from 'components/shared/TabooCard'
import { StyledPostRound } from './style'
import { completeRound } from 'store/actions/roundActions'
import { changeCardStatus } from 'store/actions/cardActions'

export class PostRound extends React.Component {
	constructor(props) {
		super(props)

		//stores the index of selected card: corresponds to the firestore deck object property of the card.
		this.state = {
			correctSelection: '',
			skippedSelection: '',
			discardSelection: '',
		}
	}

	componentDidUpdate(prevProps) {
		//If isPending.cards was true due to updating card, and now update is complete,
		//clear the selection state so buttons for that card will return to disabled
		if (prevProps.isPending.cards && !this.props.isPending.cards) {
			this.setState({
				correctSelection: '',
				skippedSelection: '',
				discardSelection: '',
			})
		}
	}
	//Called onchange of radio button value to select  card. Set in state so that on click to change the status, can determine which word is selected
	handleCardSelection = (cardIndex, status) => {
		const statuses = ['correct', 'skipped', 'discard']
		//array of the statuses that are not selected. Used to set them to empty string so only one word is selected at a time
		const notSelectedStatuses = statuses.filter((cardStatus) => cardStatus !== status)
		const property = status + 'Selection'
		//Set the card index value for the status currently selected and clear the 2 statuses not currently selected
		this.setState({
			[property]: cardIndex,
			[notSelectedStatuses[0] + 'Selection']: '',
			[notSelectedStatuses[1] + 'Selection']: '',
		})
	}

	//Called on click of the correct, skip, or discard buttons inside of taboocard.
	updateSelectedCard = (previousStatus, newStatus) => {
		const property = previousStatus + 'Selection'
		const indexToChange = this.state[property]
		//called prop method to trigger dispatch. Changes card status in firestore
		this.props.changeCardStatus(newStatus, indexToChange)
	}

	//Called when Watcher selects 'Confirm' button
	confirmRoundEnd = () => {
		//Updates score in firestore based on status of cards and game rules. Once done, updates round half and updates turn as needed.
		this.props.completeRound()
	}

	render() {
		const selections = [this.state.correctSelection, this.state.skippedSelection, this.state.discardSelection]
		const cardStatuses = ['correct', 'skipped', 'discard']
		//Creates array of the selected cards string for each status.

		//Watcher card includes buttons to updates card statuses and trigger next round

		return (
			<StyledPostRound>
				{this.props.role === 'watcher' ? (
					<PostRoundWatcher
						cardStatuses={cardStatuses}
						cardsPlayed={this.props.cardsPlayed}
						handleCardSelection={this.handleCardSelection}
						updateSelectedCard={this.updateSelectedCard}
						selectedCards={selections} // array of selected values
						confirmRoundEnd={this.confirmRoundEnd}
						isPending={this.props.isPending}
						error={this.props.error}
					/>
				) : (
					<PostRoundNonWatcher>
						{cardStatuses.map((status) => (
							<FilteredTabooList
								key={status}
								unfilteredList={this.props.cardsPlayed}
								filterKey="status"
								filterValue={status}
								displayProperty="word"
								listTitle={status}
								noneMessage={`No ${status} cards this round`}
							/>
						))}
					</PostRoundNonWatcher>
				)}
			</StyledPostRound>
		)
	}
}

PostRound.propTypes = {
	gamecode: PropTypes.string.isRequired,
	cardsPlayed: PropTypes.array.isRequired,
	role: PropTypes.string.isRequired,
	isPending: PropTypes.object.isRequired,
	error: PropTypes.object.isRequired,
	changeCardStatus: PropTypes.func.isRequired,
	completeRound: PropTypes.func.isRequired,
}

const mapStateToProps = (state) => {
	return {
		isPending: {
			round: state.round.pending,
			cards: state.cards.pending,
		},
		error: {
			round: state.round.error && state.round.error.errorMessage,
			cards: state.cards.error && state.cards.error.errorMessage,
		},
	}
}
const mapDispatchToProps = (dispatch, prevProps) => {
	const { gamecode, round, half } = prevProps
	return {
		changeCardStatus: (cardStatus, cardIndex) =>
			dispatch(changeCardStatus(gamecode, cardStatus, cardIndex, 'postround')),
		completeRound: () => dispatch(completeRound(gamecode, round, half)),
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(PostRound)
