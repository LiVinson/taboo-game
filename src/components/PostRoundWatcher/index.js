import PostRoundWatcher from "./PostRoundWatcher"

export default PostRoundWatcher