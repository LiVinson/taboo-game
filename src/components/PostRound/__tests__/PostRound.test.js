import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { ThemeProvider } from 'styled-components'
import theme from '../../../global-design/theme'
import { PostRound } from '../PostRound'
import PostRoundNonWatcher from 'components/PostRoundNonWatcher'
import PostRoundWatcher from 'components/PostRoundWatcher'

describe('PostRound renders and functions correctly', () => {
	const defaultProps = {
		gamecode: '12345',
		role: 'watcher',
		cardsPlayed: [
			{
				index: 0,
				status: 'correct',
				tabooList: ['rapper', 'West Coast', 'Compton', 'Humble', 'Good Kid'],
				word: 'Kendrick Lamar',
				roundPlayed: '1-top',
			},
			{
				index: 1,
				status: 'skipped',
				tabooList: ['Lion King', 'Disney', 'Mufasa', 'Pride Rock', 'Nala'],
				word: 'Simba',
				roundPlayed: '1-top',
			},
			{
				index: 2,
				status: 'discard',
				tabooList: ['Fresh Prince', 'Philadelphia', 'Jazzy Jeff', 'Independence Day', 'Jada Pinkett'],
				word: 'Will Smith',
				roundPlayed: '1-top',
			},
		],
		isPending: {
			round: false,
			cards: false,
		},
		error: {
			round: null,
			cards: null,
		},
		changeCardStatus: jest.fn(),
		completeRound: jest.fn(),
	}

	test('PostRound renders correctly when role is watcher', () => {
		const props = {
			...defaultProps,
		}

		const wrapper = renderer
			.create(
				<ThemeProvider theme={theme}>
					<PostRound {...props} />
				</ThemeProvider>
			)
			.toJSON()

		expect(wrapper).toMatchSnapshot()
	})
	test('PostRound renders correctly when role is not watcher', () => {
		const props = {
			...defaultProps,
			role: 'giver',
		}


		const wrapper = renderer
			.create(
				<ThemeProvider theme={theme}>
					<PostRound {...props} />
				</ThemeProvider>
			)
			.toJSON()

		expect(wrapper).toMatchSnapshot()
	})

	test('renders PostRoundWatcher when role is watcher', () => {
		const props = {
			...defaultProps,
		}

		const wrapper = shallow(
			<ThemeProvider theme={theme}>
				<PostRound {...props} />
			</ThemeProvider>
		)
		const component = wrapper.dive().find(PostRound)
		expect(component.dive().find(PostRoundWatcher)).toHaveLength(1)
		expect(component.dive().find(PostRoundNonWatcher)).toHaveLength(0)
	})

	test('renders PostRoundNonWatcher when role is not watcher', () => {
		const props = {
			...defaultProps,
			role: 'giver',
		}

		const wrapper = shallow(
			<ThemeProvider theme={theme}>
				<PostRound {...props} />
			</ThemeProvider>
		)
		const component = wrapper.dive().find(PostRound)
		expect(component.dive().find(PostRoundWatcher)).toHaveLength(0)
		expect(component.dive().find(PostRoundNonWatcher)).toHaveLength(1)
	})

	//---------Method tests --------//

	test('updateSelectedCard calls props.changeCardStatus with new status and card index', () => {
		const props = {
			...defaultProps,
		}
		const wrapper = shallow(<PostRound {...props} />)
		wrapper.instance().updateSelectedCard()
		expect(props.changeCardStatus).toHaveBeenCalled()
	})
})
