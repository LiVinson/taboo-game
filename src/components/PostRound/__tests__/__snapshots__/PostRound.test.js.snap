// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostRound renders and functions correctly PostRound renders correctly when role is not watcher 1`] = `
<div
  className="sc-fznJRM jfTSqx"
>
  <div
    className="sc-AxjAm hDlFTS"
  >
    <div
      className="sc-AxiKw dGnLMo"
    >
      Cards Played
    </div>
    <div
      className="sc-AxhCb JZaCe"
    >
      <p
        className="sc-AxheI fpDqsA"
      >
        correct
      </p>
      <ul
        className="sc-Axmtr dHjpTi"
      >
        <li
          className="sc-AxgMl fFkiZt"
        >
          Kendrick Lamar
        </li>
      </ul>
      <p
        className="sc-AxheI fpDqsA"
      >
        skipped
      </p>
      <ul
        className="sc-Axmtr dHjpTi"
      >
        <li
          className="sc-AxgMl fFkiZt"
        >
          Simba
        </li>
      </ul>
      <p
        className="sc-AxheI fpDqsA"
      >
        discard
      </p>
      <ul
        className="sc-Axmtr dHjpTi"
      >
        <li
          className="sc-AxgMl fFkiZt"
        >
          Will Smith
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`PostRound renders and functions correctly PostRound renders correctly when role is watcher 1`] = `
<div
  className="sc-fznJRM jfTSqx"
>
  <p
    className="sc-fzokOt itUVmo"
  >
    Review the cards from this round below. Select the word and click the button to change it's status. Select 
    <span
      className="sc-fznxsB gTdnEB"
    >
      Confirm!
    </span>
     when done to calculate scores.
  </p>
  <div
    className="sc-AxjAm hDlFTS"
  >
    <div
      className="sc-AxiKw dGnLMo"
    >
      correct
    </div>
    <div
      className="sc-AxhCb JZaCe"
    >
      <ul
        className="sc-Axmtr dHjpTi"
      >
        <li
          className="sc-fzqBZW iVohld"
        >
          <input
            checked={false}
            className="sc-fzqNJr iGpCVv"
            id="Kendrick Lamar"
            name="correct"
            onChange={[Function]}
            type="radio"
            value={0}
          />
          <label
            className="sc-fzoyAV iZMDEE"
            htmlFor="Kendrick Lamar"
          >
            Kendrick Lamar
          </label>
        </li>
      </ul>
    </div>
    <div
      className="sc-fzplWN buDbqD"
    >
      <button
        className="sc-AxmLO gmmVt button"
        disabled={true}
        onClick={[Function]}
      >
        skipped
      </button>
      <button
        className="sc-AxmLO sc-fzozJi jlLOkV button"
        disabled={true}
        onClick={[Function]}
      >
        discard
      </button>
    </div>
  </div>
  <div
    className="sc-AxjAm hDlFTS"
  >
    <div
      className="sc-AxiKw dGnLMo"
    >
      skipped
    </div>
    <div
      className="sc-AxhCb JZaCe"
    >
      <ul
        className="sc-Axmtr dHjpTi"
      >
        <li
          className="sc-fzqBZW iVohld"
        >
          <input
            checked={false}
            className="sc-fzqNJr iGpCVv"
            id="Simba"
            name="skipped"
            onChange={[Function]}
            type="radio"
            value={1}
          />
          <label
            className="sc-fzoyAV iZMDEE"
            htmlFor="Simba"
          >
            Simba
          </label>
        </li>
      </ul>
    </div>
    <div
      className="sc-fzplWN buDbqD"
    >
      <button
        className="sc-AxmLO gmmVt button"
        disabled={true}
        onClick={[Function]}
      >
        correct
      </button>
      <button
        className="sc-AxmLO sc-fzozJi jlLOkV button"
        disabled={true}
        onClick={[Function]}
      >
        discard
      </button>
    </div>
  </div>
  <div
    className="sc-AxjAm hDlFTS"
  >
    <div
      className="sc-AxiKw dGnLMo"
    >
      discard
    </div>
    <div
      className="sc-AxhCb JZaCe"
    >
      <ul
        className="sc-Axmtr dHjpTi"
      >
        <li
          className="sc-fzqBZW iVohld"
        >
          <input
            checked={false}
            className="sc-fzqNJr iGpCVv"
            id="Will Smith"
            name="discard"
            onChange={[Function]}
            type="radio"
            value={2}
          />
          <label
            className="sc-fzoyAV iZMDEE"
            htmlFor="Will Smith"
          >
            Will Smith
          </label>
        </li>
      </ul>
    </div>
    <div
      className="sc-fzplWN buDbqD"
    >
      <button
        className="sc-AxmLO gmmVt button"
        disabled={true}
        onClick={[Function]}
      >
        correct
      </button>
      <button
        className="sc-AxmLO sc-fzozJi jlLOkV button"
        disabled={true}
        onClick={[Function]}
      >
        skipped
      </button>
    </div>
  </div>
  <button
    className="sc-AxmLO sc-fzoLsD cyAtQH"
    disabled={false}
    onClick={[Function]}
  >
    Confirm!
  </button>
</div>
`;
